python/felis/db/database_context.py

```diff
@@ -0,0 +1,859 @@
+                for table in self.metadata.tables.values():
+                    # Fetch all existing indexes for this table once
+                    existing_indexes = _get_existing_indexes(inspector, table.name, self.schema_name)
+
+                    for index in table.indexes:
+                        if index.name is None:
+                            # Anonymous indexes can't be checked by name
+                            logger.warning(f"Skipping anonymous index on table '{table.name}'")
+                            continue
+
+                        if index.name not in existing_indexes:
+                            logger.warning(f"Skipping index '{index.name}' which does not exist")
+                            continue
+
+                        index.drop(bind=conn, checkfirst=False)  # We already checked
+                        logger.info(f"Dropped index '{index.name}'")
+            except SQLAlchemyError as e:
+                raise DatabaseContextError("Error dropping indexes", e)
+
+    def _required_schema_name(self) -> str:
+        """Return the schema name, ensuring that it is set.
+
+        This is mainly here for typing purposes, because the schema_name
+        property may be None, and mypy doesn't understand that we already
+        checked it during initialization.
+        """
+        if self.schema_name is None:
+            raise DatabaseContextError("Schema name is required but not set.")
+        return self.schema_name
+
+
+class SupportedDialect(Enum):
+    """Enumeration of supported database dialects."""
+
+    POSTGRESQL = "postgresql"
+    MYSQL = "mysql"
+    SQLITE = "sqlite"
+
+    @classmethod
+    def from_string(cls, dialect_name: str) -> SupportedDialect:
+        """Create a SupportedDialect from a string, with normalization.
+
+        Parameters
+        ----------
+        dialect_name
+            The dialect name to convert (e.g., "postgresql+psycopg2").
+
+        Returns
+        -------
+        SupportedDialect
+            The corresponding dialect enum value.
+
+        Raises
+        ------
+        ValueError
+            If the dialect is not supported.
+        """
+        # Normalize dialect name (e.g., "postgresql+psycopg2" -> "postgresql")
+        if "+" in dialect_name:
+            dialect_name = dialect_name.split("+")[0]
+
+        dialect_name = dialect_name.lower()
+
+        for dialect in cls:
+            if dialect.value == dialect_name:
+                return dialect
+
+        raise ValueError(f"Unsupported dialect: {dialect_name}")
+
+
+_ContextClass: TypeAlias = type[_BaseContext]
+_ContextDecorator: TypeAlias = Callable[[_ContextClass], _ContextClass]
+
+
+class DatabaseContextFactory:
+    """Factory for creating DatabaseContext instances based on dialect type."""
+
+    _registry: dict[SupportedDialect, _ContextClass] = {}
+
+    @classmethod
+    def register(cls, dialect: SupportedDialect) -> _ContextDecorator:
```

Comment:
I think the idea here is that the registry will only be populated at module import time and should be read-only after (thread-safety) so perhaps I would document that here.